ocaml/xapi/helpers.ml

```diff
@@ -1804,5 +1782,79 @@ end = struct
       raise e
 end
 
-let is_pool_secret_valid pool_secret =
-  List.exists (SecretString.equal pool_secret) !Xapi_globs.pool_secrets
+module PoolSecret : sig
+  val make : unit -> SecretString.t
+
+  val is_authorized : SecretString.t -> bool
+
+  val refresh_cache_or_create_new : unit -> unit
+end = struct
+  let to_secret x =
+    (* an opaque failure arises when a pool secret
+       contains a newline char, because a newline is
+       not encoded properly when putting the pool secret
+       into an http cookie. we enforce that the pool
+       secret has a certain form to avoid this kind
+       of issue
+    *)
+    let has_valid_chars =
+      Astring.(
+        String.for_all
+          (fun c -> Char.Ascii.is_alphanum c || c = '-' || c = '/')
+          x)
+    in
+    let sufficiently_secret = String.length x > 36 in
+    if has_valid_chars && sufficiently_secret |> not then
+      raise
+        Api_errors.(
+          Server_error
+            ( internal_error
+            , [
+                {|expected pool secret to match the following regex '^[0-9a-f\/\-]{37,}$'|}
+              ] )) ;
+    SecretString.of_string x
+
+  let _make () =
+    match !Xapi_globs.pool_secret_gen_method with
+    | `script ->
+        call_script !Xapi_globs.gen_pool_secret_script []
+    | `urandom ->
+        (* uuids require 16 bytes, and the pool secret is
+           made up of 3 of them, so we need 16 * 3 bytes *)
```

Comment:
Can the uuid module from transitional be used here? It already can generate uuids from urandom, we can always use its internals to use uuidm if need be.